text/0080-datetime-extension.md

```diff
@@ -0,0 +1,484 @@
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1,7].contains(context.now.dayOfWeek)
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day == 29 &&
+  principal.birthDate.month == 2 &&
+  context.now.day == 29 && 
+  context.now.month == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.now.timestamp > datetime("2020-01-31T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to millisecond accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogous to the current decimal and IP extensions), allowing users to opt-out.
+
+### Instants of Time (`datetime`)
+
+The `datetime(string)` function constructs a datetime value. Like with other extension function constructors, strict validation requires `string` to be a string literal, although evaluation/authorization support any string-typed expression. The string must be of one of the forms, and regardless of the timezone offset is always normalized to UTC:
+
+- `"YYYY-MM-DD"` (date only)
+- `"YYYY-MM-DDThh:mm:ssZ"` (UTC)
+- `"YYYY-MM-DDThh:mm:ss.SSSZ"` (UTC with millisecond precision)
+- `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (With timezone offset in hours and minutes)
+- `"YYYY-MM-DDThh:mm:ss.SSS(+/-)hhmm"` (With timezone offset in hours and minutes and millisecond precision)
+
+The `datetime` type does not provide a way for a policy author to create a `datetime` from a numeric timestamp. One of the readable formats listed above must be used, instead.
+
+Values of type `datetime` have the following methods:
+
+- `.offset(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can also be used with the `-` operator.
+
+- `DT1 - DT2` returns the difference between `DT1` and `DT2` as a `duration`. (The inverse of `-` is `DT1.offset(duration)`)
```

Comment:
Proposal: Let's not allow arithmetic operations on `DT` objects but define a function instead, e.g.
```
DT1.durationUntil(DT2)
```
to return a duration that is equivalent to DT2 - DT1. The reasons I like this approach:
- Normally, the minus operator takes two objects of the same type and returns an object of that type, e.g., two `Long`s and returns a `Long`, or two `Decimal`s and returns a `Decimal`. It's strange to break that structure when overloading minus here, at least to me. It's equally strange to support `-` but not also `+`, `*` etc. That's a signal that it's not the right operator. Creating a fresh function addresses all of these issues.
- It's perfectly clear, with the proposed approach, that what's returned is a `duration`.
- This approach still allows negative `duration` objects, i.e., when DT2 is earlier than DT1.

Reply:
Strongly agree with Mike about the `-` operator.  It’s not consistent with the usual meaning or type signature of `-`.  And I also don’t like the asymmetry of overloading just one arithmetic operator.

Let’s make this a method.

Reply:
I like `durationBetween`.

Reply:
`durationBetween` doesn't make it very clear in which order the subtraction happens.  I'd rather we had `.since` or `durationSince` .

Reply:
In my mind, `DT1.durationUntil(DT2)` maps to `DT2 - DT1`, `DT1.durationSince(DT2)` maps to `DT1 - DT2`, and `DT1.durationBetween(DT2)` maps to the absolute value of the difference.

Examples:
* `context.now.durationUntil(principal.startDate) > 0` - I'd expect the principal hasn't started yet
* `context.now.durationSince(principal.startDate) > 0` - I'd expect the principal already started
* `context.now.durationBetween(principal.startDate) > 0` - could be either

Given this, I think my preference is for `DT1.durationSince(DT2)` since that keeps things in the same order as `DT1 - DT2`. I'd also be open to supporting both `durationSince` and `durationUntil`.

Reply:
@khieta I agree, now that you point it out, that between implies absolute duration which is less desirable (at least in my opinion). :+1: for `durationSince`.

Reply:
`durationSince` works for me. I prefer just having that, and not also having `durationUntil`, for simplicity and to hopefully avoid confusion.